### PR TITLE
Add support for bzlmod

### PR DESCRIPTION
**Pull request by mortenmj**
This includes the commits from #132. I'd prefer we merged that first and then rebased this PR.
I haven't figured out how to avoid the dependency on rules_go/gazelle if running binaries (setting dev_dependency=True breaks things), so we'd still force people to load those, which is unfortunate.

For a repository consuming the snapshots rules, the MODULE.bazel would look like this, when running with the snaptool built from source:

```
bazel_dep(name = "com_cognitedata_bazel_snapshots", version = "0.0.0")
```

and with binaries:

```
bazel_dep(name = "com_cognitedata_bazel_snapshots", version = "0.0.0")

snapshots = use_extension("@com_cognitedata_bazel_snapshots//snapshots:extensions.bzl", "snapshots")
snapshots.toolchains(from_source = False)

use_repo(snapshots, "snapshots_snaptool_toolchains")

register_toolchains("@snapshots_snaptool_toolchains//:all")
```

Those last four lines are only needed to override the default registration in the MODULE.bazel in the snapshots repo.

### CODE REVIEW

**Comment by mortenmj** [comment]
We should update the workspace snippet before merging this, to also include bzlmod instructions. Similar to e.g.

https://github.com/bazelbuild/rules_python/releases/tag/0.20.0

File: MODULE.bazel [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module(
+    name = "com_cognitedata_bazel_snapshots",
+    version = "0.6.2"
```
Comment (psalaberria002) [inline MODULE.bazel:3]: How is this field updated on new releases? Manually?

File: MODULE.bazel [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+)
+
```
Comment (psalaberria002) [inline MODULE.bazel:5]: renovate doesn't support bzlmod yet, I saw someone started to work on a proposal tho.

File: MODULE.bazel [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+bazel_dep(name = "rules_go", version = "0.39.0", repo_name = "io_bazel_rules_go")
+bazel_dep(name = "gazelle", version = "0.30.0", repo_name = "bazel_gazelle")
+bazel_dep(name = "bazel_skylib", version = "1.4.1")
+
+snapshots = use_extension("@com_cognitedata_bazel_snapshots//snapshots:extensions.bzl", "snapshots")
+snapshots.toolchains(from_source = True)
+
+use_repo(
```
Comment (psalaberria002) [inline MODULE.bazel:13]: Can gazelle update this list, or manual for now?

File: examples/use-binaries/WORKSPACE [hunk 4]
```diff
@@ -1,4 +1,4 @@
-workspace(name = "use-binaries")
+workspace(name = "build-from-source")
```
Comment (psalaberria002) [inline examples/use-binaries/WORKSPACE:1]: The example folder is called use-binaries. Is this rename coming from a search/replace command?

Comment (mortenmj) [reply examples/use-binaries/WORKSPACE:1]: Yes. Fixing

File: snapshots/extensions.bzl [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+# buildifier: disable=unused-variable
+def _snapshots_extension(module_ctx):
+    go_dependencies()
+
+    registrations = {}
+    for mod in module_ctx.modules:
+        for toolchains in mod.tags.toolchains:
+            if toolchains.name != "snapshots" and not mod.is_root:
+                fail("""\
+                Only the root module may override the default name for the oci toolchains.
```
Comment (psalaberria002) [inline snapshots/extensions.bzl:27]: copy paste

File: snapshots/extensions.bzl [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                This prevents conflicting registrations in the global namespace of external repos.
+                """)
+
+            if toolchains.name not in registrations:
+                registrations[toolchains.name] = toolchains.from_source
+
+    for name, from_source in registrations.items():
+        snapshots_register_toolchains(name, from_source = from_source, register = False)
```
Comment (psalaberria002) [inline snapshots/extensions.bzl:35]: Why do we hardocde register = False?

Comment (psalaberria002) [reply snapshots/extensions.bzl:35]: Nevermind, I see the comment in the function args now.

File: MODULE.bazel (hunk 1 above)
Comment (mortenmj) [reply MODULE.bazel:3]: I suggest we also leave out the bzlmod snippet from the release documentation until we have a better story for how to import from other repos, unless we want it to be something like

```
bazel_dep(name = "com_cognitedata_bazel_snapshots", version = "0.7.0")
archive_override(
  module_name = "com_cognitedata_bazel_snapshots",
  urls = ["https://github.com/cognitedata/bazel-snapshots/releases/download/0.7.0/bazel-snapshots-0.7.0.tar.gz"],
  integrity = "sha goes here",
  strip_prefix = "bazel-snapshots-0.7.0",
)
```

That's doable, but not very neat

Comment (psalaberria002) [reply MODULE.bazel:3]: Agree.

File: snapshots/extensions.bzl (hunk 5 above)
Comment (mortenmj) [reply snapshots/extensions.bzl]: Fixed

**Pull request merged**